aten/src/ATen/native/transformers/hip/flash_attn/flash_api.hip

```diff
@@ -0,0 +1,651 @@
+  CHECK_SHAPE(v, batch_size, seqlen_k, num_heads_k, head_size_og);
+
+  at::Tensor q_padded, k_padded, v_padded;
+  q_padded = q;
+  k_padded = k;
+  v_padded = v;
+
+  at::Tensor out;
+  if (out_.has_value()) {
+    out = out_.value();
+    TORCH_CHECK(out.dtype() == q_dtype, "Output must have the same dtype as inputs");
+    CHECK_DEVICE(out);
+    TORCH_CHECK(out.stride(-1) == 1, "Output tensor must have contiguous last dimension");
+    CHECK_SHAPE(out, batch_size, seqlen_q, num_heads, head_size_og);
+    if (head_size_og % 8 != 0) { out = at::empty_like(q_padded); }
+  } else {
+    out = at::empty_like(q_padded);
+  }
+
+  auto round_multiple = [](int x, int m) { return (x + m - 1) / m * m; };
+  const int head_size = round_multiple(head_size_og, 8);
+  const int head_size_rounded = round_multiple(head_size, 32);
+  const int seqlen_q_rounded = round_multiple(seqlen_q, 128);
+  const int seqlen_k_rounded = round_multiple(seqlen_k, 128);
+
+  // Otherwise the kernel will be launched from cuda:0 device
+  // Cast to char to avoid compiler warning about narrowing
+  at::hip::HIPGuardMasqueradingAsCUDA device_guard{(char)q.get_device()};
+
+  // We want to checkpoint and save the RNG state for backward if dropout
+  // We get the default generator and return the seed and offset which will
+  // be used in the backward function
+  auto gen = at::get_generator_or_default<at::CUDAGeneratorImpl>(c10::nullopt, at::cuda::detail::getDefaultCUDAGenerator());
+  at::Tensor seed_t, offset_t;
+
+  if (p_dropout > 0.0)  {
+    // number of times random will be generated per thread, to offset philox counter in thc random
+    // state
+    // We use a custom RNG that increases the offset by batch_size * nheads * 32.
+    int64_t counter_offset = batch_size * num_heads * 32;
+    // See Note [Acquire lock when using random generators]
+    std::lock_guard<std::mutex> lock(gen->mutex_);
+    at::PhiloxCudaState philox_state = gen->philox_cuda_state(counter_offset);
+    if (at::cuda::currentStreamCaptureStatus() == at::cuda::CaptureStatus::None) {
+      auto [seed, offset] = at::cuda::philox::unpack(philox_state);
+      seed_t = at::scalar_tensor(at::Scalar(static_cast<int64_t>(seed)), at::dtype(at::kLong));
+      offset_t = at::scalar_tensor(at::Scalar(static_cast<int64_t>(offset)), at::dtype(at::kLong));
+    } else {
+      seed_t = at::empty({}, at::dtype(at::kLong).device(at::kCUDA));
+      offset_t = at::empty({}, at::dtype(at::kLong).device(at::kCUDA));
+    }
+  } else {
+    if (at::cuda::currentStreamCaptureStatus() != at::cuda::CaptureStatus::None) {
+      seed_t = at::empty({}, at::dtype(at::kLong).device(at::kCUDA));
+      offset_t = at::empty({}, at::dtype(at::kLong).device(at::kCUDA));
+    } else {
+      seed_t = at::empty({}, at::dtype(at::kLong));
+      offset_t = at::empty({}, at::dtype(at::kLong));
+    }
+
+  }
+
+  auto stream = at::hip::getCurrentHIPStreamMasqueradingAsCUDA().stream();
+  //reorder tensors and make contiguous
```

Comment:
yeah if you are making them contiguous you likely don't want the stride(-1) check above

Reply:
Technically our AOT Triton kernel doesn't require them to be continuous, but we haven't tested them yet.